### PR TITLE
Fix logic of exclude branches

### PR DESCRIPTION
**Pull request by imbstack**
This adds some tests and changes the logic of exclude branches a bit. [Bug 1433506](https://bugzilla.mozilla.org/show_bug.cgi?id=1433506)

### CODE REVIEW

File: src/intree.js [hunk 1]
```diff
@@ -128,15 +128,38 @@ module.exports.setup = function(cfg) {
           return false;
         }
 
+        let event = payload.details['event.type'];
         let events = task.task.extra.github.events;
         let branches = task.task.extra.github.branches;
         let branch = payload.details['event.base.repo.branch'];
```
Comment (djmitche) [inline src/intree.js]: can we call this `includeBranches` for parallelism

File: src/intree.js [hunk 2]
```diff
@@ -128,15 +128,38 @@ module.exports.setup = function(cfg) {
...
+
+          let included = branches && _.includes(branches, branch);
+          let excluded = excludeBranches && _.includes(excludeBranches, branch);
+
+          if (included && excluded) {
+            throw new Error(`Cannot include and exclude same branch. '${branch}' is in both lists!`);
+          }
+
+          if (branches && excludeBranches) {
+            return included && !excluded;
```
Comment (djmitche) [inline src/intree.js]: isn't this option not allowed now?

**Review by djmitche — CHANGES_REQUESTED** [review]

**Comment by staktrace** [comment]
@djmitche Looks like the patch was updated, could you re-review? Thanks!

File: src/intree.js [hunk 3]
```diff
@@ -128,15 +128,36 @@ module.exports.setup = function(cfg) {
...
+            return false;
+          }
+
+          if (event !== 'push') {
+            return true;
+          }
+
+          if (includeBranches && excludeBranches) {
+            throw new Error(`Cannot include and exclude same branch. '${branch}' is in both lists!`);
+          }
```
Comment (djmitche) [inline src/intree.js]: I think the error message here is incorrect.  The condition here is that there is both a `branches` and an `excludeBranches`.  The README suggests this situation is allowed (although I don't know what that would mean -- if a branch isn't on either list, is it included or excluded?).

I think when we talked about this briefly, we decided that this case is not allowed -- you must use no more than one of `branches` or `excludeBranches`.  Then there's no concern with the same branch appearing in both lists.

**Review by djmitche — CHANGES_REQUESTED** [review]

**Review by djmitche — APPROVED** [review]

**Pull request merged**